### PR TITLE
Add binder server

### PR DESCRIPTION
**Pull request by TaWeiTu**
In order for the server to listen for binder connections, the users should tell the server to listen on a "binder port" using the binder server credentials.  A "binder port" is represented as "binder://" followed by the service name, for example,

```
grpc::ServerBuilder server_builder;
server_builder.AddListeningPort("binder://echo.service", grpc::experimental::BinderServerCredentials());
```

the server will now listen for binder connections to service "echo.service". The service name can be an arbitrary string and is used to retrieve the endpoint binder as follows

```
std::unique_ptr<grpc::Server> server = server_builder.BuildAndStart();
void* endpoint_binder = grpc::experimental::binder::GetEndpointBinder("echo.service");
```

The Java service can then pass this endpoint binder to clients.

Depends on #26970.

@ctiller
@sifmelcara

### CODE REVIEW

File: src/core/ext/transport/binder/server/binder_server.h
Comment (sifmelcara) [inline src/core/ext/transport/binder/server/binder_server.h:34]: Maybe we want to add a lock?

**Review by sifmelcara — COMMENTED** [review]
Most of the code came from other pending PRs?

Will do a review again after other PRs get merged

File: src/core/ext/transport/binder/server/binder_server.h
Comment (TaWeiTu) [reply src/core/ext/transport/binder/server/binder_server.h]: Done.

**Comment by TaWeiTu** [comment]
> Most of the code came from other pending PRs?

Yeah I have to build this PR on top of the other one in order to make the tests work. The relevant files in this PR are:
* `binder_server.h`, `binder_server.cc`
* `binder_server_credentials.h`, `binder_server_credentials.cc`
* `binder_server_test.cc`

> Will do a review again after other PRs get merged

Sure, no problem!

File: src/core/ext/transport/binder/server/binder_server_credentials.cc
Comment (sifmelcara) [inline src/core/ext/transport/binder/server/binder_server_credentials.cc:29]: Need to include corresponding header, otherwise

src/core/ext/transport/binder/server/binder_server_credentials.cc:29:24: error: no member named 'TransactionReceiverAndroid' in namespace 'grpc_binder'

Comment (TaWeiTu) [reply src/core/ext/transport/binder/server/binder_server_credentials.cc:29]: Thanks! I forgot to try building in Android.

**Comment by TaWeiTu** [comment]
Hi Craig, can you take a look at the overall API design of the binder server? It should be easier to review now since the gigantic PR it depends on has been merged. Thanks!

**Comment by TaWeiTu** [comment]
> TaWei: Any idea why `Android (Internal CI) ` build suddenly cannot find NDK headers used in our headers?

I guess we were not building any Android code before in the CI since it's not included in tests, nor is it a target in `BUILD` and `CMakeLists.txt`. The Android docker only has `ndk-bundle` installed but not `ndk;21.4.7075529`, and I suppose that's the reason why the build failed. Not sure which of the following options is better:

* Install `ndk;21.4.7075529` in that docker
* Instead of determining `GPR_ANDROID` or `defined(ANDROID) || defined(__ANDROID__)`, additionally checks whether NDK binder is available.

**Comment by sifmelcara** [comment]
Ok then I think we should also check for binder API availability in the guard. Our code simply won't work with old version of NDK and API level.

I will do that

File: include/grpc/grpc.h
Comment (ctiller) [inline include/grpc/grpc.h:447]: Public API additions will need a gRFC written.

File: include/grpcpp/security/server_credentials.h
Comment (ctiller) [inline include/grpcpp/security/server_credentials.h:73]: Public API's will need a gRFC.

File: src/core/ext/transport/binder/server/binder_server.h
Comment (ctiller) [reply src/core/ext/transport/binder/server/binder_server.h:50]: @veblush what's holding up using absl::Mutex everywhere (which can be initialized statically?)

File: src/core/ext/transport/binder/server/binder_server.h
Comment (ctiller) [inline src/core/ext/transport/binder/server/binder_server.h:143]: Why not bool?

File: src/core/ext/transport/binder/server/binder_server.h
Comment (ctiller) [inline src/core/ext/transport/binder/server/binder_server.h:142]: Make this template argument an interface instead.

File: include/grpc/grpc.h
Comment (TaWeiTu) [reply include/grpc/grpc.h:447]: I wonder if we still need this C API for us to wrap C++ API around it. According to https://github.com/grpc/grpc/projects/8 it seems no longer relevant.

File: src/core/ext/transport/binder/server/binder_server.h
Comment (TaWeiTu) [reply src/core/ext/transport/binder/server/binder_server.h]: Done.

File: src/core/ext/transport/binder/server/binder_server.h
Comment (TaWeiTu) [reply src/core/ext/transport/binder/server/binder_server.h]: I've made it into a lambda that constructs the receiver.

File: include/grpcpp/security/server_credentials.h
Comment (TaWeiTu) [reply include/grpcpp/security/server_credentials.h]: We decided to temporarily move the APIs back to our own namespace `grpc_binder` because we need this PR in the internal testings and the RFC process is likely to take longer than desired. We're drafting a gRFC, and will make the APIs public once the RFC is approved. Thanks!

File: include/grpc/grpc.h
Comment (ctiller) [reply include/grpc/grpc.h]: Whilst I think it's probably a worthwhile idea, I don't think it's an actual official stance for the project (and if it were I think a gRFC would be needed for that).

We need a C API, and we need a gRFC for it.

**Comment by sifmelcara** [comment]
Public APIs has been removed. We will post a gRFC today or tomorrow. Craig: We would like to merge this first to unblock our implementation of CI testing (both public and internal), and adjust/extend our interface during the RFC review. WDYT?

File: include/grpc/grpc.h
Comment (TaWeiTu) [reply include/grpc/grpc.h]: Sure, we will include the this C API into our gRFC.

File: src/core/ext/transport/binder/server/binder_server.h
Comment (ctiller) [reply src/core/ext/transport/binder/server/binder_server.h:50]: Pinging @veblush again

**Review by ctiller — APPROVED** [review]

**Pull request merged**
